src/main/scala/org/graphframes/pattern/patterns.scala

```diff
@@ -48,13 +49,124 @@ private[graphframes] object PatternParser extends RegexParsers {
 private[graphframes] object Pattern {
   def parse(s: String): Seq[Pattern] = {
     import PatternParser._
-    parseAll(patterns, s) match {
+    val result = parseAll(patterns, s) match {
       case result: Success[_] =>
         result.asInstanceOf[Success[Seq[Pattern]]].get
       case result: NoSuccess =>
         throw new InvalidParseException(
           s"Failed to parse bad motif string: '$s'.  Returned message: ${result.msg}")
     }
+    assertValidPatterns(result)
+    result
+  }
+
+  /**
+   * Checks all Patterns for validity:
+   *  - Disallow named edges within negated terms
+   *  - Disallow term "()-[]->()" and its negation
+   *  - Disallow name to be shared by a vertex and an edge
+   * @throws InvalidParseException if an negated terms contain named edges
+   */
+  private def assertValidPatterns(patterns: Seq[Pattern]): Unit = {
+
+    // vertexNames, edgeNames are used to check for duplicate names across vertices and edges
+    val vertexNames = mutable.HashSet.empty[String]
+    val edgeNames = mutable.HashSet.empty[String]
+    def addVertex(v: Vertex): Unit = v match {
+      case NamedVertex(name) =>
+        if (edgeNames.contains(name)) {
+          throw new InvalidParseException(s"Motif reused name '$name' for both a vertex and " +
+            s"an edge, which is not allowed.")
+        }
+        vertexNames += name
+      case AnonymousVertex =>  // pass
+    }
+    def addEdge(e: Edge): Unit = e match {
+      case NamedEdge(name, src, dst) =>
+        if (vertexNames.contains(name)) {
+          throw new InvalidParseException(s"Motif reused name '$name' for both a vertex and " +
+            s"an edge, which is not allowed.")
+        }
+        edgeNames += name
+        addVertex(src)
+        addVertex(dst)
+      case AnonymousEdge(src, dst) =>
+        addVertex(src)
+        addVertex(dst)
+    }
+
+    patterns.foreach {
```

Comment:
This method looks like it could be combined with the addVertex, addEdge logic above, but it's awkward to since we need to handle Negation differently.